### DISCUSSION TITLE
[BUG]: How do i import react native video in swift file (iOS native side)

### DISCUSSION THREAD

**Discussion by venky145** [TheWidlarzGroup/react-native-video#3564] [Ideas]
### What problem are you trying to solve?

I'm attempting to access the AVPlayer instance from RCTVideo in my local Swift file in order to retrieve some parameters from the AVPlayer. However, I'm facing difficulties in importing the react-native-video module into my Swift file and accessing the AVPlayer instance.

**Actual Behavior:**
I'm unable to import react-native-video into my Swift file and consequently unable to access the AVPlayer instance.

### Proposed solution

Be able to import the react-native-video module (`import RCTVideo`) into a local Swift file and access the AVPlayer instance from RCTVideo without any issues.

### Platforms

iOS

### Version line

v6

### Additional context

**Architecture:** Old architecture

**Steps to Reproduce:**
1. Attempt to import react-native-video into a local Swift file.
2. Try to access the AVPlayer instance from RCTVideo in the Swift file.

```swift
import RCTVideo
```

**Expected Behavior:**
I expect to be able to import the react-native-video module into my Swift file and access the AVPlayer instance without any issues.

**Reproduction:** Unable to import RCTVideo to my own swift file.

**Comment by KrzysztofMoch** [comment]
Hmm... can you try to use `modular_headers = true` ?

**Comment by venky145** [comment]
I have tried this also before but it did not work
  pod 'react-native-video', path: '../node_modules/react-native-video', :modular_headers => true


My pod file

```
require_relative '../node_modules/react-native/scripts/react_native_pods'
require_relative '../node_modules/@react-native-community/cli-platform-ios/native_modules'
require_relative '../node_modules/react-native-permissions/scripts/setup'

platform :ios, min_ios_version_supported
prepare_react_native_project!

# ⬇️ uncomment wanted permissions
setup_permissions([
   'AppTrackingTransparency',
  # 'BluetoothPeripheral',
  # 'Calendars',
  # 'Camera',
  # 'Contacts',
  # 'FaceID',
  # 'LocationAccuracy',
  # 'LocationAlways',
  # 'LocationWhenInUse',
  # 'MediaLibrary',
  # 'Microphone',
  # 'Motion',
   'Notifications',
  # 'PhotoLibrary',
  # 'PhotoLibraryAddOnly',
  # 'Reminders',
  # 'Siri',
  # 'SpeechRecognition',
  # 'StoreKit',
])

# If you are using a `react-native-flipper` your iOS build will fail when `NO_FLIPPER=1` is set.
# because `react-native-flipper` depends on (FlipperKit,...) that will be excluded
#
# To fix this you can also exclude `react-native-flipper` using a `react-native.config.js`
# ```js
# module.exports = {
#   dependencies: {
#     ...(process.env.NO_FLIPPER ? { 'react-native-flipper': { platforms: { ios: null } } } : {}),
# ```
#flipper_config = ENV['NO_FLIPPER'] == "1" ? FlipperConfiguration.disabled : FlipperConfiguration.enabled
flipper_config = FlipperConfiguration.disabled

linkage = ENV['USE_FRAMEWORKS']
if linkage != nil
  Pod::UI.puts "Configuring Pod with #{linkage}ally linked Frameworks".green
  use_frameworks! :linkage => linkage.to_sym
end

target 'fluxion' do
  use_modular_headers!
  
  config = use_native_modules!
  $RNVideoUseGoogleIMA = true

  # Flags change depending on the env values.
  flags = get_default_flags()

  use_react_native!(
    :path => config[:reactNativePath],
    # Hermes is now enabled by default. Disable by setting this flag to false.
    # Upcoming versions of React Native may rely on get_default_flags(), but
    # we make it explicit here to aid in the React Native upgrade process.
    :hermes_enabled => false,
    :fabric_enabled => flags[:fabric_enabled],
    # Enables Flipper.
    #
    # Note that if you have use_frameworks! enabled, Flipper will not work and
    # you should disable the next line.
    :flipper_configuration => flipper_config,
    # An absolute path to your application root.
    :app_path => "#{Pod::Config.instance.installation_root}/.."
  )
  
  pod 'react-native-video', path: '../node_modules/react-native-video', :modular_headers => true


  target 'fluxionTests' do
    inherit! :complete
    # Pods for testing
  end

  post_install do |installer|
    react_native_post_install(
      installer,
      # Set `mac_catalyst_enabled` to `true` in order to apply patches
      # necessary for Mac Catalyst builds
      :mac_catalyst_enabled => false
    )
    __apply_Xcode_12_5_M1_post_install_workaround(installer)
#  react_native_post_install(installer)
  installer.pods_project.targets.each do |target|
     target.build_configurations.each do |config|
         config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] ||= ['$(inherited)', '_LIBCPP_ENABLE_CXX17_REMOVED_UNARY_BINARY_FUNCTION']
         if config.build_settings['IPHONEOS_DEPLOYMENT_TARGET'].to_f < 12.0
            config.build_settings['IPHONEOS_DEPLOYMENT_TARGET'] = '12.0'
          end
      end
  end
  installer.pods_project.targets.each do |target|
    if target.respond_to?(:product_type) and target.product_type == "com.apple.product-type.bundle"
      target.build_configurations.each do |config|
          config.build_settings['CODE_SIGNING_ALLOWED'] = 'NO'
      end
    end
  end
  end
end
```

**Comment by KrzysztofMoch** [comment]
Do you have podspec for your native module ?

**Comment by venky145** [comment]
@KrzysztofMoch 
Here is is my podspec
[react-native-video.podspec.zip](https://github.com/react-native-video/react-native-video/files/14485541/react-native-video.podspec.zip)

**Comment by venky145** [comment]
Just to update

I am able to import as like below in my swift class and can access RCTVideo making it as public

import react_native_video


private var _video:RCTVideo?

**Comment by freeboub** [comment]
@venky145 I will have to do something similar very soon
The goal is to integrate an analytic tool which listen to AVPlayer events.
my advise:
- you need to create a singleton on swift code, it shall be pubic and usable for anywhere in swift
- In the singleton, you need one function to register all RCTVideo instance, each new instance will register on mount and unregister on umount
- In the singleton, you need to create a function registerAnalyticPlugin(AnalyticPluginInterface).
- this AnalyticPluginInterface need to define 2 functions, onPlayerIntanceCreated(AVPlayer) onPlayerIntanceDestroyed(AVPlayer)


Than in your plugin, you need to access to the singleton and registerAnalyticPlugin(YourAnalyticPluginInterface).
once playback starts, it will call onPlayerIntanceCreated with the player instance

Do it make sens on your side ?
Notice I already implement this on android, it works fine

**Comment by freeboub** [comment]
@venky145 I just create a plugin system here: https://github.com/TheWidlarzGroup/react-native-video/pull/3909
It will allow to access to AVPlayer handle from a side package !

**Comment by KrzysztofMoch** [comment]
plugins system should resolve this - I will close this
